packages/api/src/external/carequality/patient.ts

```diff
@@ -1,48 +1,86 @@
+import { createPatientDiscoveryRequest } from "./create-pd-request";
+import { makeIheGatewayAPI } from "./api";
+
+dayjs.extend(duration);
+
+const createContext = "cq.patient.discover";
+export const PATIENT_DISCOVERY_TIMEOUT = dayjs.duration({ minutes: 2 });
+const CHECK_DB_INTERVAL = dayjs.duration({ seconds: 5 });
+type RaceControl = { isRaceInProgress: boolean };
 
 export async function discover(
   patient: Patient,
   organization: Organization,
   facilityNPI: string
 ): Promise<void> {
-  let iheGateway: IHEGateway | undefined;
   try {
-    console.log(`IHE Gateway patient discovery - M patientId ${patient.id}`);
-
-    iheGateway = makeIheGatewayAPI();
+    const { id, cxId } = patient;
+    const iheGateway = makeIheGatewayAPI();
+    console.log(`Kicking off patient discovery for patientId: ${id}`);
 
     const fhirPatient = toFHIR(patient);
 
-    const iheGatewayRequest = patientToIheGateway({
+    const nearbyCQOrgs = await searchNearbyCQOrganizations({
+      cxId,
+      patientId: id,
+    });
+    const xcpdGateways = cqOrgsToXCPDGateways(nearbyCQOrgs);
+
+    const pdRequest = createPatientDiscoveryRequest({
       patient: fhirPatient,
       cxId: patient.cxId,
-      xcpdGateways: [], // TODO ADD XCPD GATEWAYS WHEN WE HAVE THEM
-      facilityNPI: facilityNPI,
+      xcpdGateways,
+      facilityNPI,
       orgName: organization.data.name,
       orgOid: organization.oid,
     });
 
-    iheGateway.startPatientDiscovery(iheGatewayRequest);
+    // Intentionally asynchronous - we will be checking for the results in the database
+    iheGateway.startPatientDiscovery([pdRequest]);
+
+    const raceControl: RaceControl = { isRaceInProgress: true };
+    // Run the patient discovery until it either times out, or all the results are in the database
+    const raceResult = await Promise.race([
+      startTimeout(),
+      checkNumberOfResults(raceControl, pdRequest.id, pdRequest.xcpdGateways.length), // TODO: RAISE A FOLLOW UP - set up an event listener for Mirth
+    ]);
+    if (raceResult) {
+      console.log(`${raceResult}. Starting to handle patient discovery results. PatientId: ${id}`);
+      raceControl.isRaceInProgress = false;
+    }
+
+    // At this point, discovery results are stored in the database, so we can retrieve them
+    const discoveryResults = await getPatientDiscoveryResults(pdRequest.id);
+    await handlePatientDiscoveryResults(patient, discoveryResults);
+
+    analytics({
+      distinctId: cxId,
+      event: EventTypes.patientDiscovery,
+      properties: {
+        apiType: Product.medical,
+        numberGateways: xcpdGateways.length,
+        numberLinkedGateways: discoveryResults.length,
+      },
+    });
   } catch (err) {
-    const msg = `Failure while starting patient discovery for ${patient.id} @ IHE Gateway`;
+    const msg = `Failed to carry out patient discovery for ${patient.id}`;
     console.error(msg, err);
     capture.message(msg, {
```

Comment:
I know it was already there, but now we can use `capture.error(msg, ...)` - where `msg` is a string`, so we don't need to set `level: "error"`.

Reply:
It was, until a couple of days ago. Things move fast around here! 😅

Reply:
🦔 🦸